### PR TITLE
[serverless] implement an adaptive flush automatically deciding the best moment to flush

### PR DESCRIPTION
**Pull request by remeh**
### What does this PR do?

In the Serverless Agent / Datadog Lambda Extension, this PR adds the computing of the frequency of execution of the current lambda function. From this, it is capable of selecting the best flush strategy: flush at the end of the invocation, flush at the start of the invocation, or flush at least every N seconds (at the start of an invocation).

The main purpose being to avoid consuming extra CPU time during the execution of a lambda function.

### Additional Notes

Use `DD_SERVERLESS_DISABLE_ADAPTIVE_FLUSH` to always use the "flush at the end of the invocation" strategy.

The behavior is:
  - if a function is invoked more than once a minute, flush at the start of an invocation at least every 10 second 
  - else, if a function is executed more than once every 5 minute, flush at the start of the function invocations,
  - else, flush at the end of the function invocations.

This PR also implements timeout flush cancellation of the Logs Agent synchronous flush using `context.Context`.

### CODE REVIEW

File: pkg/serverless/frequency.go
Comment (DarcyRaynerDD) [reply pkg/serverless/frequency.go:24]: Alternatively, we could use an [exponential moving average](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average). That way we don't allocate any memory, and we weight for the most recent invocations. It's usually pretty simple, eg `newAverage = lastAverage + averagingFactor * ( newValue - lastAverage)`. The averaging factor is a constant we pick, usually 0<x<1. The larger it is, the more weight recent invocations have.

File: pkg/serverless/frequency.go
Comment (DarcyRaynerDD) [inline pkg/serverless/frequency.go:33]: nit: Maybe InvocationInterval? I don't think this a frequency.

File: pkg/serverless/frequency.go
Comment (DarcyRaynerDD) [inline pkg/serverless/frequency.go:9]: 50 seems like a lot of invocations to look back. Worried about it not swapping strategies fast enough when there is a change in traffic

File: pkg/serverless/flush/strategy.go
Comment (DarcyRaynerDD) [inline pkg/serverless/flush/strategy.go:84]: Seems like we don't use this strategy. Maybe we should delete.

**Review by DarcyRaynerDD — COMMENTED** [review]
Really nice work. Left some nits and general thoughts

File: pkg/serverless/flush/strategy.go
Comment (nhinsch) [inline pkg/serverless/flush/strategy.go:27]: Is `Running` supposed to be commented out here?

File: pkg/serverless/flush/strategy_test.go
Comment (nhinsch) [inline pkg/serverless/flush/strategy_test.go:9]: I would add some tests for the other strategies as well

File: pkg/serverless/frequency.go
Comment (nhinsch) [inline pkg/serverless/frequency.go:53]: Seconding Darcy's comment, I think this is more of an interval (length of time) rather than a frequency (rate per amount of time)

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go:27]: May reappear with further changes but indeed, it is unused in this first implementation. Removing it 👍

File: pkg/serverless/flush/strategy_test.go
Comment (remeh) [reply pkg/serverless/flush/strategy_test.go:9]: Very true, adding some.

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go:84]: Will do.

File: pkg/serverless/frequency.go
Comment (remeh) [reply pkg/serverless/frequency.go:53]: Correct.

File: pkg/serverless/frequency.go
Comment (remeh) [reply pkg/serverless/frequency.go:24]: About the memory usage, for completeness, with 50 values, we are speaking of `24 + 50*8=424 bytes`.
About the lookback, I don't mind reducing its size since I feel like the adaptive flush will mainly be useful for function quickly reaching their interval of invocation.

First I'll push a change with less values for now, let's say 10, and see in a second time if switching to a moving average is better, that's a good idea.

File: pkg/serverless/frequency.go
Comment (remeh) [reply pkg/serverless/frequency.go:9]: Changing it to `10`.

File: pkg/serverless/flush/strategy_test.go
Comment (remeh) [reply pkg/serverless/flush/strategy_test.go]: Done.

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go]: Renamed to `interval` and it is now private. The reason it was public was to be able to build an object outside of this package, as simple as this object was, I preferred that to a constructor. But I don't mind having a constructor so I've added one.

File: cmd/serverless/main.go
Comment (hush-hush) [inline cmd/serverless/main.go:240]: Why not use the config package ?

File: pkg/logs/processor/processor.go
Comment (hush-hush) [inline pkg/logs/processor/processor.go:68]: What's the difference with  return ?

File: pkg/logs/processor/processor.go
Comment (hush-hush) [inline pkg/logs/processor/processor.go:65]: This we're using a `default we could simply do:

```go
if ctx.Err() != nil {
    return
}
```

And remove the switch. I thinks it's clearer and make it explicit that the `default` is blocking and can't be canceled. WDYT ?

File: pkg/logs/sender/batch_strategy.go
Comment (hush-hush) [inline pkg/logs/sender/batch_strategy.go:47]: Same here for `ctx.Err()`

File: pkg/serverless/flush/strategy.go
Comment (hush-hush) [inline pkg/serverless/flush/strategy.go:24]:
```suggestion
	// Starting will trigger a flush at the start of the function
```

File: pkg/serverless/flush/strategy.go
Comment (hush-hush) [inline pkg/serverless/flush/strategy.go:27]:
```suggestion
	// Stopping will trigger a flush at the end of the function
```

File: pkg/serverless/flush/strategy.go
Comment (hush-hush) [inline pkg/serverless/flush/strategy.go:20]: Why a string and not an int ?

File: pkg/serverless/flush/strategy.go
Comment (hush-hush) [inline pkg/serverless/flush/strategy.go:1]: We're missing the copyright

File: pkg/serverless/flush/strategy.go
Comment (hush-hush) [inline pkg/serverless/flush/strategy.go:25]: What about `Started` since starting implies a short time at the start of the function but `Periodically` could flush minutes after this ?

File: pkg/serverless/interval.go
Comment (hush-hush) [inline pkg/serverless/interval.go:1]: Missing copyright here too

File: pkg/serverless/interval.go
Comment (hush-hush) [inline pkg/serverless/interval.go:65]: Could you explain why ? What is the gain of flushing 10s after the start of the function if I run every 3 minutes instead of 5 ?

File: pkg/serverless/flush/strategy.go
Comment (hush-hush) [inline pkg/serverless/flush/strategy.go:40]: We should use `NewPeriodically`

File: pkg/serverless/protocol.go
Comment (hush-hush) [inline pkg/serverless/protocol.go:66]: What's the point ?

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/serverless/interval.go
Comment (remeh) [reply pkg/serverless/interval.go:65]: I think you've mixed up what we are trying to achieve.
Basically, the idea is to flush at the start of the function invocation instead of at the end, because it means that the flush will run in parallel with the function execution. But we can't do that for every function (hence the run interval detection) because there is scenario where we would never flush / discard some data if we would do so.

Here, we enable the "periodical" flush only with functions running often enough. We're not flushing 10s after the start of the function, we're flushing "at least every 10 seconds". Meaning, if the function is executed every 500ms for instance, the function would run 20 times in 10s, but we would do only one flush, at the start of the 20th invocation, avoiding to do too many useless flush (cpu time) and not losing any data. Let me know if that is still not clear, I can also share the RFC with you if needed.

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go:25]: It's the proper term, the function is just starting.

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go:20]: If used in a debug log or anything, it's way more readable (even while  `%+v` for the developer), and since there like 1 or 2 instances of this at runtime, we don't mind about memory usage.

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go:27]: It's not correct, `Stopping` is used as a flag for the adaptive flush to know that it has been queried right after the function is stopping (and it won't/could not trigger a flush).

File: pkg/serverless/flush/strategy.go
Comment (remeh) [reply pkg/serverless/flush/strategy.go:24]: It's not correct, Starting is used as a flag for the adaptive flush to know that it has been queried while the function is just starting (and it won't/could not trigger a flush).

File: pkg/logs/processor/processor.go
Comment (remeh) [reply pkg/logs/processor/processor.go:65]: This `select` + `case <-ctx.Done()` is a well known pattern while using `context.Context` for cancellation (even the one documented as far as I recall but not entirely sure anymore).

File: cmd/serverless/main.go
Comment (remeh) [reply cmd/serverless/main.go:240]: It is not yet declared in the `config` package (this one and some others from the initial implementation) 🙇 
However that is definitely the goal before GA, I'm waiting for this PR to be merged: https://github.com/DataDog/datadog-agent/pull/7313 which starts using the `config` package to also switch to using config + supporting configuration file.

File: pkg/serverless/interval.go
Comment (hush-hush) [reply pkg/serverless/interval.go:70]: Ok, Thanks for the explanation.

Still why 5min ? Is it arbitrary ?

Comment (remeh) [reply pkg/serverless/interval.go:70]: It is based on timings measured by the Serverless team and with a few assumptions made together, so not exactly arbitrary even if a bit of a guess work for now.

Comment (hush-hush) [reply pkg/serverless/interval.go:70]: OK, thanks for the update. Do we want to add a comment explaining the reasoning ?

**Review by hush-hush — APPROVED** [review]
Approved, feel free to add a comment about the 5 min threshold if you think it would be useful.

File: pkg/serverless/interval.go
Comment (remeh) [reply pkg/serverless/interval.go:70]: Yeah good idea, I'm on it.

File: pkg/serverless/interval.go
Comment (DarcyRaynerDD) [inline pkg/serverless/interval.go:22]: nit:
```suggestion
	// ignore points older than the last stored one
```

**Review by DarcyRaynerDD — APPROVED** [review]
LGTM, only had one nit

**Pull request merged**
